### ISSUE TITLE
Determining input equivalence with File checksums and Directory listings

### ISSUE THREAD

**Issue by jdidion** [openwdl/wdl#472]
A shortcoming of the `File` and `Directory` data types is the inability to guarantee the equivalence of inputs, which breaks reproducibility and makes the implementation of call caching (which we call "job reuse" on DNAnexus) difficult. Say I have the following directory in my cloud storage:

```
foo
|_bar
|_baz
```

I run a job with a `Directory`-type input and provide `foo` as the value. Now I add a new file `blorf` to the `foo` directory, and I replace `baz` with a new file of the same name but with different contents. I run the job again. Given just the directory name, how does my implementation know that the contents of that directory have remained unchanged since the first time I ran the job?

I propose to formally define two alternative JSON formats for the `File` and `Directory` types in the standard [input/output formats](https://github.com/openwdl/wdl/blob/main/versions/development/SPEC.md#input-and-output-formats). Rather than reinvent the wheel, we will just borrow the terminology from CWL.

```
{
  "file1": "/path/to/file",
  "dir1": "/path/to/dir",
  "dir2": {
    "location": "/path/to/dir",
    "listing": [
      {
        "type": "File",
        "basename": "foo.txt",
        "checksum": "sha256:ABC123"
      },
      {
        "type": "Directory",
        "basename": "bar"
        "listing": [
          {
            "type": "File",
            "basename": "baz.txt",
            "checksum": "md5:WTFBBQ42"
          }
        ]
      }
    ]
  },
  "dir3": {
    "listing": [
      {
        "type": "File",
        "location": "/path/to/foo.txt",
        "basename": "bar.txt",
        "checksum": "sha256:ABC123"
      },
    ]
  }
}
```

In the simple form, a `File`/`Directory` value is just a string - typically a local path or a URI. The object forms have the following fields:

* `File`
    * `type`: Always "File"; optional at the top-level but required within directory listings
    *  `location`: The file location - this is equivalent to the value in the simple form. May be absent if the file is within a listing as long as `basename` is specified.
    * `basename`: The name of the file relative to the containing directory. If the basename differs from the actual file name at the given location, the file must be localized with the given basename.
    * `checksum`: A checksum of the file using one of the approved algorithms. If specified, the checksum must be verified during localization.
* `Directory`
    * `type`: Always "Directory"; optional at the top-level but required within directory listings
    * `location`: The directory location - this is equivalent to the value in the simple form. May be absent if the directory is within a listing as long as `basename` is specified.
    * `basename`: The name of the directory relative to the containing directory. If the basename differs from the actual directory name at the given location, the file must be localized with the given basename.
    * `listing`: An array of files/subdirectories within the directory. May be nested to any degree.

Importantly, none of these fields will be exposed within WDL, so the runtime definition of `File`/`Directory` won't change.

**Comment by patmagee** [comment]
This is a great idea @jdidion and is something I have been wondering about myself.  One quick question I have here, would you consider exposing some of those properties within wdl proper. A File might only ever be assigned with a url, but once defined should it have attributes (checksum, basename, size even?).

**Comment by mlin** [comment]
Hi all, just a comment here, I'd consider generalizing/weakening this to say that the File/Directory representation may be a JSON object with a `location` key and *whatever else* the engine may wish to include or interpret. Going beyond that may be too prescriptive of implementation details. For example, miniwdl's call cache just uses the filesystem mtimes instead of digests.

**Comment by jdidion** [comment]
@mlin Good point. Checksum shouldn't be required. But we can make a suggestion to use checksums or some other means of determining file equality. And to Ruben's point we can suggest checksum algorithms, but require that any specific algorithm be used.

**Comment by jdidion** [comment]
Partially addressed in WDL 1.2. I think the open question is whether to make recommendations in the spec around checksums.
